app/controllers/components_controller.rb

```diff
@@ -2,15 +2,19 @@
 
 class ComponentsController < ApplicationController
   def show
-    @component = Component.get(params[:id])
+    @component = Component.get(params[:id]) || not_found
   end
 
   def index
     @components = Component.all
   end
 
   def preview
-    @component = Component.get(params[:component_id])
+    @component = Component.get(params[:component_id]) || not_found
     render layout: 'preview'
   end
+
+  def not_found
+    render status: 404, text: "404 error"
```

Comment:
Ahh okay. It looks like if you just do `render status: :not_found` it will try and render the show view with a 404 status code, of course the view code is unsupported so you then get an 500 error. By adding the text body you are avoiding that which makes sense.

You don't actually need to return a body with the response for the error handling functionality further up the application stack to work so you can simply do this:

```rb
def show
  @component = Component.get(params[:id])

  head :not_found if @component.nil?
end
```